### PR TITLE
Update example_ec2 to Create Instance Which Supports Hibernation (#35790)

### PR DESCRIPTION
**Pull request by dashton90**
A bug was introduced in the ec2_example DAG when adding the `EC2HibernateInstanceOperator`. This PR updates the configuration of the example to create an instance which supports hibernation.

Discussion begins [here](https://github.com/apache/airflow/pull/35790#issuecomment-1825004349).
 
related: #35790

### CODE REVIEW

**Comment by dashton90** [comment]
@ferruzzi tagging you since you flagged the bug

**Comment by ferruzzi** [comment]
At a glance, it looks right.  As mentioned, we'll need to adjust it to accept a KMS key to encrypt the EBS volume.  Github makes it a bit of a pain to collaborate on a PR like this; it would be nice to be able to add a commit to your PR myself, but as far as I'm aware I can't do that so..... here's what needs to be done next:
 

Replace Line 41 with the following block:

```
KMS_KEY_ID_KEY = "KMS_KEY_ID"

sys_test_context_task = (
    SystemTestContextBuilder()
    .add_variable(KMS_KEY_ID_KEY)
    .build()
)
```

then add a line between 97 and 98:

```
kms_key_id = test_context[KMS_KEY_ID_KEY]
```

Then you can use that in the config{}.  I'm not positive on this part, but I /think/ you apply it by changing line 113 to 

```
{"DeviceName": "/dev/xvda", "Ebs": {"Encrypted": True, "KmsKeyId": kms_key_id, "DeleteOnTermination": True}}
```

I would need to test that last part out, but the first few steps are definitely right and will allow us to store the KMS key in SSM Parameter Store and fetch it for the test.  

Check out the [RDS Export test](https://github.com/apache/airflow/blob/main/tests/system/providers/amazon/aws/example_rds_export.py) for a working example of a KMS Key in a system test if I didn't make sense.

Unfortunately, as you mentioned, once you do that it still needs one of us to actually test running it against the real service using the real KMS key.  But if we merge this as it is, we'll need to make yet another PR to make those changes.... I'm not sure which is actually more convenient.

**Comment by dashton90** [comment]
I added you as a collaborator to my fork so you should be free to push any changes you need. But I can make the changes myself tonight or tomorrow if you don't get around to it.

I skipped creating the KMS key because Amazon will default to using an Amazon managed key to encrypt the volume if you don't specify one. Figured it was less overhead than managing our own key, but happy to create one if that's the established practice.

**Comment by ferruzzi** [comment]
>Amazon will default to using an Amazon managed key to encrypt the volume if you don't specify one.

AH! I misunderstood, I thought you meant "Amazon" meaning my team specifically would have to add it.   If the API call works without specifying one, then by all means let's skip that whole thing and keep the example/test as simple as we can.

**Comment by ferruzzi** [comment]
Alright.   progress!   I ran the proposed changes and it failed, but it's getting much further.  Here's the current state:

```
INFO     airflow.task.operators.airflow.providers.amazon.aws.operators.ec2.EC2TerminateInstanceOperator:ec2.py:249 Terminating EC2 instance i-03e2f168f51570fec
ERROR    airflow.models.taskinstance.TaskInstance:taskinstance.py:2696 Task failed with exception
Traceback (most recent call last):
  File "/opt/airflow/airflow/models/taskinstance.py", line 433, in _execute_task
    result = execute_callable(context=context, **execute_callable_kwargs)
  File "/opt/airflow/airflow/providers/amazon/aws/operators/ec2.py", line 251, in execute
    ec2_hook.get_waiter("instance_terminated").wait(
  File "/usr/local/lib/python3.8/site-packages/botocore/waiter.py", line 55, in wait
    Waiter.wait(self, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/botocore/waiter.py", line 375, in wait
    raise WaiterError(
botocore.exceptions.WaiterError: Waiter InstanceTerminated failed: Waiter encountered a terminal failure state: For expression "Reservations[].Instances[].State.Name" we matched expected path: "stopping" at least once
```

**Comment by dashton90** [comment]
Huh, it looks like boto [fails the terminate waiter if the instance is in the stopping state](https://github.com/boto/botocore/blob/81c0e57059de3ed7a56bf183e322a83f7a9e80ec/botocore/data/ec2/2014-10-01/waiters-2.json#L230-L235).

No idea if that is a bug or the intended behaviour since instances which are stopping should be able to be terminated [per the AWS docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html)

Easiest fix is probably to add `wait_for_completion=True` to both the Reboot and Hibernate operators. I can do that now

**Comment by ferruzzi** [comment]
In response to my last comment, it looks like the consensus is to get this sorted out, then we (I) can split it into two tests and just run them in parallel.   We're fine with the run time, but breaking it out might save a bit of time.  I'm running into unrelated issues at the moment with Docker, but I'm poking at it as I can.

**Comment by ferruzzi** [comment]
Alright, I dialed it in.   It passed with `max_retries` set to 75 and fails at 60, so let's go with 75 for now.   We can bump it later as needed, but that seems like a reasonable starting point.

**Comment by dashton90** [comment]
Done in https://github.com/apache/airflow/pull/35839/commits/00e06ce86601f8c2c54b9e16ccb6177477c76b7a

**Comment by ferruzzi** [comment]
I merged main into the branch and that got the CI tests to pass.  :rocket: 

The only other thing I'd like to see is move the wait_for_completion and the max_retries out of the code snippets as I mentioned above so they don't clutter up the docs pages, but other than that I'm very happy with it.

Can we get another pair of eyes?  @o-nikolas @vincbeck @syedahsn

**Review by syedahsn — APPROVED** [review]

File: tests/system/providers/amazon/aws/example_ec2.py [hunk 1]
```diff
@@ -50,11 +50,14 @@ def get_latest_ami_id():
...
     image_prefix = "Amazon Linux*"
+    root_device_name = "/dev/xvda"
 
     images = boto3.client("ec2").describe_images(
         Filters=[
             {"Name": "description", "Values": [image_prefix]},
-            {"Name": "architecture", "Values": ["arm64"]},
+            {"Name": "architecture", "Values": ["x86_64"]},
+            {"Name": "root-device-type", "Values": ["ebs"]},
+            {"Name": "root-device-name", "Values": [root_device_name]},
```
Comment (o-nikolas) [inline tests/system/providers/amazon/aws/example_ec2.py]: Why do we need these changes?

File: tests/system/providers/amazon/aws/example_ec2.py [hunk 2]
```diff
@@ -149,13 +155,16 @@ def parse_response(instance_ids: list):
     reboot_instance = EC2RebootInstanceOperator(
         task_id="reboot_instace",
         instance_ids=instance_id,
+        wait_for_completion=True,
```
Comment (o-nikolas) [inline tests/system/providers/amazon/aws/example_ec2.py]: Move this and below out of the docs snippet.

Comment (dashton90) [reply tests/system/providers/amazon/aws/example_ec2.py]: Done in https://github.com/apache/airflow/pull/35839/commits/b8e65688f4d6db438465c82d2146333d2e117e71

File: tests/system/providers/amazon/aws/example_ec2.py (hunk 1 above)
Comment (dashton90) [reply tests/system/providers/amazon/aws/example_ec2.py]: Non-graviton instances only support x86 architecture. The root device must be ebs backed in order to support hibernation

**Comment by dashton90** [comment]
Moved everything out of the codeblock in https://github.com/apache/airflow/pull/35839/commits/b8e65688f4d6db438465c82d2146333d2e117e71

File: tests/system/providers/amazon/aws/example_ec2.py (hunk 1 above)
Comment (ferruzzi) [reply tests/system/providers/amazon/aws/example_ec2.py]: Hibernation requires an x86 architecture and encrypted storage; encrypted storage requires EBS; EBS requires manually assigning mount points.  It was a bit of a chain of requirements there.

**Review by ferruzzi — APPROVED** [review]
LGTM.  Your persistence is greatly appreciated on this one, it was a bit of a slog.

File: tests/system/providers/amazon/aws/example_ec2.py (hunk 1 above)
Comment (vincbeck) [reply tests/system/providers/amazon/aws/example_ec2.py]: Could we add comments? That'd be helpful I think :)

File: tests/system/providers/amazon/aws/example_ec2.py (hunk 2 above)
Comment (o-nikolas) [reply tests/system/providers/amazon/aws/example_ec2.py]: Thanks!

File: tests/system/providers/amazon/aws/example_ec2.py (hunk 1 above)
Comment (o-nikolas) [reply tests/system/providers/amazon/aws/example_ec2.py]: Thanks, that's helpful context. I agree with @vincbeck, I'd love some comments here explaining that for future folks.

Comment (dashton90) [reply tests/system/providers/amazon/aws/example_ec2.py]: Added comments in https://github.com/apache/airflow/pull/35839/commits/4089321bc32e1f5c21246afc3920e8b804aed0a6

**Review by vincbeck — APPROVED** [review]

File: tests/system/providers/amazon/aws/example_ec2.py (hunk 1 above)
Comment (vincbeck) [reply tests/system/providers/amazon/aws/example_ec2.py]: Thanks a lot @dashton90

**Comment by vincbeck** [comment]
Congrats @dashton90 , thanks to your PR, the system test is now succeeding, see dashboard [here](https://aws-mwaa.github.io/#/open-source/system-tests/dashboard.html)

**Pull request merged**